### PR TITLE
Add adaptive majority voting for GRPO training

### PR DESCRIPTION
**Pull request by AfraAmini**
## Description
This PR introduces a majority-aware objective for GRPO training. When the `--aggregate_maj` flag is enabled, the advantage computation is modified to account for majority responses. The key change involves accumulating ratios across all samples for each prompt, following the formula:

$$\frac{1}{K}\sum_{k=1}^K A(y_k) (\prod_{k'=1}^K w(y_k'))$$

where:
- $K$ is the number of samples per prompt
- $A(y_k)$ is the majority advantage for the k-th sample
- $w(y_k)$ is the importance weight (exp(log prob diff)) for the k-th sample

## Changes Made
- Added new function `compute_majority_reward_per_prompt` in `model_utils.py` to calculate rewards based on majority answers
- Modified data preparation pipeline:
  - Removed random shuffling to preserve sample ordering
  - Retained samples with zero advantage (they contribute to ratio calculations)
  - Disabled packing to maintain sample relationships
- Moved the `backward()` call (line 985 in `grpo_fast`) outside the loop

The data preparation and optimization loop modifications are where things get very inefficient, and I appreciate any help figuring out a better way to implement it.

### CODE REVIEW

File: open_instruct/grpo_fast.py [hunk 1]
```diff
@@ -951,12 +976,34 @@ def train(
...
+                    pg_losses = -single_advantages * sum_ratios
+                    pg_losses2 = -single_advantages * torch.clamp(sum_ratios, 1.0 - args.cliprange, 1.0 + args.cliprange)
+                    pg_loss_max = torch.max(pg_losses, pg_losses2)
+                    loss_maj = pg_loss_max.mean()                    
+                    # Clear any existing gradients before backward pass
+                    self.model.zero_grad()
+                    # Check if loss is valid before backward pass
+                    if not torch.isnan(loss_maj) and not torch.isinf(loss_maj) and loss_maj != 0:
+                        self.model.backward(loss_maj)
+                        self.model.step()
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py]: how do you handle gradient accumulation here?

File: open_instruct/grpo_fast.py [hunk 2]
```diff
@@ -951,12 +976,34 @@ def train(
...
+                        # )
+                        # pg_loss_stats[i] = masked_mean(pg_loss_max, mb_response_masks_bool, args.masked_mean_axis)
+                        # loss_stats[i] = loss if not args.aggregate_maj else 0.0
+                        # ratio_stats[i] = masked_mean(ratio, mb_response_masks_bool, args.masked_mean_axis)
+                if args.aggregate_maj:    
+                    logprobs_diffs = torch.stack(logprob_diffs).reshape(-1, args.num_samples_per_prompt_rollout)
+                    
+
+                    sum_ratios = torch.exp(torch.sum(logprobs_diffs, dim=-1)).unsqueeze(-1)
+                    single_advantages = torch.stack(single_advantages).reshape(-1, args.num_samples_per_prompt_rollout)
```
Comment (hamishivi) [inline open_instruct/grpo_fast.py]: I worry a bit about the stacking. Maybe pre allocate and fill, try torch.cat, or removing some of the stacks?

**Review by hamishivi — COMMENTED** [review]
Two things popped out to me from a quick look

**Comment by finbarrtimbers** [comment]
Closing due to inactivity. Feel free to reopen!
